src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -129,48 +149,52 @@ public async ValueTask<FlushResult> WriteAsync(SerializedHubMessage hubMessage,
         // TODO: We could consider buffering messages until they hit backpressure in the case when the connection is down
         await _resend.Task.ConfigureAwait(false);
 
+        ValueTask<FlushResult> writeTask;
```

Comment:
```suggestion
        // Await writeTask outside of the _writeLock so AckAsync can make forward progress.
        ValueTask<FlushResult> writeTask;
```

Feel free to improve the comment, but this was a key part to avoiding the deadlock I assume?